### PR TITLE
Implement getSnapshotProperties() "inherit from same-named snapshot" concept

### PR DESCRIPTION
**Pull request by jimklimov**
This follows up from #384 and #506 to hopefully avoid a potential issue with "inherit" mode ignoring the parent's `dst_X_synced` values. 

Testing privately, review welcome to clear for merging. I hope it won't eat our data :)

### CODE REVIEW

**Comment by jimklimov** [comment]
From testing:
* znapzending a dataset `nvpool/SHARED` which is a policy root:
````
[Mon Aug  3 17:12:55 2020] [debug] sending snapshots from nvpool/SHARED/opt/glassfish-4.0 to backup-adata/snapshots/nvpool/SHARED/opt/glassfish-4.0
# zfs list -H -o name -t snapshot -s creation -d 1 nvpool/SHARED/opt/glassfish-4.0
# zfs list -H -o name -t snapshot -s creation -d 1 backup-adata/snapshots/nvpool/SHARED/opt/glassfish-4.0
# zfs send -Lce -i 'nvpool/SHARED/opt/glassfish-4.0@znapzend-auto-2020-08-03T14:56:40Z' 'nvpool/SHARED/opt/glassfish-4.0@znapzend-auto-2020-08-03T15:08:18Z'|/usr/bin/amd64/mbuffer -q -s 256k -W 600 -m 1G|zfs recv -u -F 'backup-adata/snapshots/nvpool/SHARED/opt/glassfish-4.0'
# zfs list -H -o name -t snapshot nvpool/SHARED/opt/glassfish-4.0@znapzend-auto-2020-08-03T15:08:18Z
# zfs set org.znapzend:dst_0_synced=1 nvpool/SHARED/opt/glassfish-4.0@znapzend-auto-2020-08-03T15:08:18Z
# zfs set org.znapzend:dst_0=backup-adata/snapshots/nvpool/SHARED/opt/glassfish-4.0 nvpool/SHARED/opt/glassfish-4.0@znapzend-auto-2020-08-03T15:08:18Z
...
[Mon Aug  3 17:17:32 2020] [debug] got nothing to clean in children of source nvpool/SHARED
# zfs list -H -o name -t snapshot -s creation -d 1 nvpool/SHARED/opt/glassfish-4.0
# zfs list -H -o name -t snapshot -s creation -d 1 nvpool/SHARED/opt/glassfish-4.0
# zfs list -H -o name -t snapshot -s creation -d 1 backup-adata/snapshots/nvpool/SHARED/opt/glassfish-4.0
[Mon Aug  3 17:17:34 2020] [debug] not cleaning up source nvpool/SHARED/opt/glassfish-4.0@znapzend-auto-2020-08-03T15:08:18Z because it is needed by backup-adata/snapshots/nvpool/SHARED/opt/glassfish-4.0
[Mon Aug  3 17:17:34 2020] [debug] got nothing to clean in children of source nvpool/SHARED
...
````
From the above I see that it:
** successfully sets `dst_X_synced=1` property,
** honors this property by not cleaning the just-created snapshot,
** ...which it should not have cleaned anyway, so I suppose it should have gone looking to keep the older second-newest snapshot? or at least not inspected/logged this one?
** logs many times (for each child and their children) the verbatim same mantra "got nothing to clean in children of source nvpool/SHARED" (TODO: check the loop and message) UPDATE: fixed

**Comment by jimklimov** [comment]
For an inherited mode:

````
[Mon Aug  3 19:41:44 2020] [debug] now will look if there is anything to clean in children of destination backup-adata/snapshots/nvpool/SHARED/var/adm
# zfs list -H -o name -t snapshot -s creation -d 1 nvpool/SHARED/var/adm
# zfs list -H -o name -t snapshot -s creation -d 1 nvpool/SHARED/var/adm
# zfs get -H -s local,inherited -t snapshot -r -o property,value all nvpool/SHARED/var/adm@znapzend-auto-2020-08-03T17:41:40Z

### Here it is recursing into parent (hm, should somehow make snapshotExists quieter in case of missing snapshot):

# zfs list -H -o name -t snapshot nvpool/SHARED/var@znapzend-auto-2020-08-03T17:41:40Z
cannot open 'nvpool/SHARED/var@znapzend-auto-2020-08-03T17:41:40Z': dataset does not exist

[Mon Aug  3 19:41:45 2020] [debug] not cleaning up source nvpool/SHARED/var/adm@znapzend-auto-2020-08-03T17:41:40Z recursively because it is needed by backup-adata/snapshots/nvpool/SHARED/var/adm
[Mon Aug  3 19:41:45 2020] [debug] cleaning up source snapshots recursively under nvpool/SHARED/var/adm
# WOULD # zfs destroy -r nvpool/SHARED/var/adm@znapzend-auto-2020-08-03T16:37:20Z,znapzend-auto-2020-08-03T16:51:50Z,znapzend-auto-2020-08-03T16:58:52Z,znapzend-auto-2020-08-03T17:27:57Z,znapzend-auto-2020-08-03T17:33:11Z,znapzend-auto-2020-08-03T17:34:49Z
[Mon Aug  3 19:41:45 2020] [debug] now will look if there is anything to clean in children of source nvpool/SHARED/var/adm
[Mon Aug  3 19:41:45 2020] [info] done with backupset nvpool/SHARED/var/adm in 5 seconds
[Mon Aug  3 19:41:45 2020] [debug] send/receive worker for nvpool/SHARED/var/adm done (17830)
````

So seems to work and try to protect the snapshots now. Usually however this fast-tracks to the older `lastAndCommonSnapshots()` when both source and destination are up, instead of looping to call the new `getSnapshotProperties()` in `mostRecentCommonSnapshot()` -- to test that, have to neuter the clause for lastAndCommonSnapshots (`if (0) {...}`).

In either case it seems to protect the newest snapshot made (which is reasonable since it is replicated after all) and can consider dropping older snapshots (e.g. if I want 1 per hour, a 3-minute old one from earlier test is not needed anymore).

I did not yet manage to get it to try and delete a snapshot made earlier in backupSet's (`nvpool/SHARED`) history, maybe its scheduled time did not come yet.

Sometimes it ends up with this error however, when looking at a missing snapshot -- probably gotta merge hashes differently:
````
# zfs list -H -o name -t snapshot nvpool@znapzend-auto-2020-08-03T17:45:53Z
cannot open 'nvpool@znapzend-auto-2020-08-03T17:45:53Z': dataset does not exist
Reference found where even-sized list expected at /export/home/jim/shared/znapzend/bin/../lib/ZnapZend/ZFS.pm line 1035.
Reference found where even-sized list expected at /export/home/jim/shared/znapzend/bin/../lib/ZnapZend/ZFS.pm line 1035.
````

And also better find a way to not spew "dataset does not exist" to stderr where we anticipate it may not exist.

**Comment by jimklimov** [comment]
Algorithmic flaws: when we do get into the mode of looking for older `dst_X_synced` property flags (dest is offline so the just-made snapshot is not the newest common known), we now iterate into parent datasets to inspect their same-named snapshots and newly check if they have the property defined (and seems it is not inheritable into same-named snapshots of child datasets).

The check can optionally be done first in recursive mode, so zfs can destroy all same-named datasets in a tree starting from current backupSet->src root, and then we follow up by inspecting snapshots that remain in child datasets to see if any of those should be cleared (e.g. had earlier run inherited mode and have obsolete snapshots that are not defined in the root backupSet).

* we should not do this iterative approach for `dst_X_synced` if the current dataset is already the real backupSet (directly has the policy defined, we are not using the "--inherited" mode), and so not waste some processing time in the default mode for most users
* we should not re-check in child datasets the remaining snapshot names we already inspected (and decided to keep alive) in their parent, our current backupSet (inherited or root)
* it seems that when we descent into child datasets and hit a branch where several sub-datasets have same snapshot names, e.g. there was an "inherited" mode run from that point in tree, the current iteration goes from each such child and grandchild up in the tree, calling `zfs get` to look for `dst_X_synced`, and so re-parses the same data over and over (mostly finding nothing). While this is pedantically correct, it can easily spiral to be too costly in I/O, forking, processing... so there should be some safe and reliable way that we can cache the results and quickly see that "from this dataset@snapname upwards we know this property is defined to X, or is definitely not-set" and only request each such data once. And somehow not eat all RAM (fall back to current mode if lowMemRecursion?)...
* we should avoid getting all snapshot/dataset properties when we know we are going to use one or two specific names and discard the rest - pass those names to `zfs get propname1,propname2,... dataset@snap` right away.

**Comment by jimklimov** [comment]
Seems we have a nice bonus for the `dst_X_synced` use-case: the code also stores `dst_X` as a property of the snapshot that was the current backupSet source (e.g. for inherited ones too), which allows us to cut iterations short when we hit the true parent of a snapname:

````
# zfs get all nvpool/SHARED/var/adm@znapzend-auto-2020-08-03T16:51:50Z | grep local
nvpool/SHARED/var/adm@znapzend-auto-2020-08-03T16:51:50Z  org.znapzend:dst_0          backup-adata/snapshots/nvpool/SHARED/var/adm                       local
nvpool/SHARED/var/adm@znapzend-auto-2020-08-03T16:51:50Z  org.znapzend:dst_0_synced   1                                                                  local

:; zfs get all nvpool/SHARED/var/adm | grep local
:; zfs get all nvpool/SHARED/var | grep local
:; zfs get all nvpool/SHARED | grep local
nvpool/SHARED  org.znapzend:dst_0_plan     1weeks=>30minutes,2months=>1days,4months=>1weeks,10years=>1months  local
nvpool/SHARED  org.znapzend:dst_0          backup-adata/snapshots/nvpool/SHARED                               local
nvpool/SHARED  org.znapzend:mbuffer_size   1G                                                                 local
nvpool/SHARED  org.znapzend:src_plan       1weeks=>30minutes,2months=>1days,4months=>1weeks,10years=>1months  local
nvpool/SHARED  org.znapzend:zend_delay     0                                                                  local
nvpool/SHARED  org.znapzend:tsformat       znapzend-auto-%Y-%m-%dT%H:%M:%SZ                                   local
nvpool/SHARED  org.znapzend:post_znap_cmd  off                                                                local
nvpool/SHARED  org.znapzend:mbuffer        /usr/bin/amd64/mbuffer                                             local
nvpool/SHARED  org.znapzend:pre_znap_cmd   off                                                                local
nvpool/SHARED  org.znapzend:recursive      on                                                                 local
nvpool/SHARED  org.znapzend:enabled        on                                                                 local

```

File: lib/ZnapZend/ZFS.pm [hunk 1]
```diff
@@ -553,7 +557,75 @@ sub sendRecvSnapshots {
...
+    #   string with our name-space propertyPrefix (e.g. "org.znapzend:...")
+    #   prepended if needed;
+    # * undef for defaulting to 'all'
+    # Returns a string safe to pass into `zfs get`
+    my $self = shift;
+    my $propnames = shift;
+
+    my $propertyPrefix = $self->propertyPrefix;
+
+    if (defined($propnames) && $propnames) {
```
Comment (oetiker) [inline lib/ZnapZend/ZFS.pm:575]: this does the same

```suggestion
    if ($propnames) {
```

File: lib/ZnapZend/ZFS.pm [hunk 2]
```diff
@@ -553,7 +557,75 @@ sub sendRecvSnapshots {
...
+        if (ref($propnames) eq 'ARRAY') {
+            if (scalar(@$propnames) > 0) {
+                my $propstring = '';
+                foreach my $propname (@$propnames) {
```
Comment (oetiker) [inline lib/ZnapZend/ZFS.pm:579]:
```suggestion
                for my $propname (@$propnames) {
```

foreach is rather oldstyle

File: lib/ZnapZend/ZFS.pm [hunk 3]
```diff
@@ -553,7 +557,75 @@ sub sendRecvSnapshots {
...
+                    next if !defined($propname);
```
Comment (oetiker) [inline lib/ZnapZend/ZFS.pm:580]:
```suggestion
                    next unless $propname;
```
since $propmane is probably never 0 or an empty string this should be fine

File: lib/ZnapZend/ZFS.pm [hunk 4]
```diff
@@ -553,7 +557,75 @@ sub sendRecvSnapshots {
...
+                    chomp $propname;
+
+                    if ($propname eq 'all') {
+                        warn "=== getSnapshotProperties(): got an 'all' propname in array, any filtering will be moot";
```
Comment (oetiker) [inline lib/ZnapZend/ZFS.pm:584]: using 'warn' can be problematic as it does not go to the log ...

File: lib/ZnapZend/ZFS.pm [hunk 5]
```diff
@@ -948,45 +1023,167 @@ sub getSnapshotProperties {
...
+                ###     `use v5.10.1;` or newer for ~~ smartmatch operator
+                ### Still warns about experimental perl feature in 5.22...
+                #if ($] >= 5.010001) {
+                #    if (@wantedPropnames ~~ \%properties) {
+                #        $inherit = 0;
+                #    }
+                #} else {
+                    # Old ways...
+                    my $uniqHits = 0;
+                    foreach my $key (keys %properties) {
```
Comment (oetiker) [inline lib/ZnapZend/ZFS.pm:1141]:
```suggestion
                    for my $key (sort keys %properties) {
```
sort keys for reproducability and use for

File: lib/ZnapZend/ZFS.pm [hunk 6]
```diff
@@ -948,45 +1023,167 @@ sub getSnapshotProperties {
...
+                        if (exists($seenPropnames{$key})) {
```
Comment (oetiker) [inline lib/ZnapZend/ZFS.pm:1142]:
```suggestion
                        if (defined $seenPropnames{$key}) {
```

File: lib/ZnapZend/ZFS.pm (hunk 3 above)
Comment (jimklimov) [reply lib/ZnapZend/ZFS.pm:585]: Would a bad caller with numeric `0` be a problem?
We have an explicit check for `eq ''` below, would that suffice if we do not check `defined` here?

Comment (oetiker) [reply lib/ZnapZend/ZFS.pm:585]: $propname with "", "0", 0 is false .. question is if that is a problem

File: lib/ZnapZend/ZFS.pm (hunk 5 above)
Comment (jimklimov) [reply lib/ZnapZend/ZFS.pm:1146]: What good would be sorting here? We anyway check all key values, so I guess it would only waste time to sort them?

File: lib/ZnapZend/ZFS.pm (hunk 4 above)
Comment (jimklimov) [reply lib/ZnapZend/ZFS.pm:589]: Note to self: I'll revise the use of `print STDERR` too - there was a lot of such copypasta :(

File: lib/ZnapZend/ZFS.pm (hunk 6 above)
Comment (jimklimov) [reply lib/ZnapZend/ZFS.pm:1147]: Is there a particular difference (performance etc?) And also `defined (stuff)` vs. `defined stuff` - are these equivalent?

File: lib/ZnapZend/ZFS.pm (hunk 3 above)
Comment (jimklimov) [reply lib/ZnapZend/ZFS.pm]: Well, this is supposed to be a sanity-checker of sorts for whatever other authors would put into their arrays eventually, maybe generated ones. Array elements that are `undef` or `""` we skip(ed) explicitly. If evolution provides a zfs property named `0` or `org.znapzend:0` we would fail to get it with this change...

File: lib/ZnapZend/ZFS.pm (hunk 1 above)
Comment (jimklimov) [reply lib/ZnapZend/ZFS.pm]: So the shorter form is true if:
* propnames is defined,
* AND propnames is not (empty string OR zero number (OR maybe string "0" too?))

In particular, if it is NOT defined, there is no dereferencing error to check for further values?

File: lib/ZnapZend/ZFS.pm (hunk 6 above)
Comment (oetiker) [reply lib/ZnapZend/ZFS.pm]: With exists you only know that the key exists ... the value might still be undefined
defined with brackets is equivalent ... that is just a style thing

File: lib/ZnapZend/ZFS.pm (hunk 3 above)
Comment (oetiker) [reply lib/ZnapZend/ZFS.pm:585]: ok

File: lib/ZnapZend/ZFS.pm (hunk 5 above)
Comment (oetiker) [reply lib/ZnapZend/ZFS.pm]: true :) I often use sort just to ensure that two runs are exactly the same ... but I guess it would not matter in this instance

File: lib/ZnapZend/ZFS.pm (hunk 3 above)
Comment (jimklimov) [reply lib/ZnapZend/ZFS.pm:585]: So in the end, not applying this one, now - consciously :)

File: lib/ZnapZend/ZFS.pm (hunk 6 above)
Comment (jimklimov) [reply lib/ZnapZend/ZFS.pm]: thanks

File: lib/ZnapZend/ZFS.pm (hunk 4 above)
Comment (jimklimov) [reply lib/ZnapZend/ZFS.pm]: Uff, that was an unexpected long adventure :)

**Comment by jimklimov** [comment]
At the moment it looks pretty neat and functional. There may be some points of improvement as listed in detail above, primarily:
* reduce unneeded I/O and processing load from requests to `zfs` command getting data we had seen recently
** stop recursing into parents when we pass a border (root backupSet, initial backupSet this run is iterating from) even if we do not find a way to cache all seen results
* reduce noise from `snapshotExists()` probing a snapshot that does not exist and we anticipate it is not there - find way to hide its STDERR into /dev/null if we asked for `$quiet==1`


but even so, performance does not seem to take extreme hits on reasonably small amounts of datasets, and when destination is online, and when we do not "runonce inherit" which is the primary mode of operation - so these and other improvements can happen in other PRs and by other authors ;)

File: lib/ZnapZend/ZFS.pm (hunk 1 above)
Comment (jimklimov) [reply lib/ZnapZend/ZFS.pm]: Question aside, the proposed fix is merged

File: lib/ZnapZend.pm [hunk 7]
```diff
@@ -535,14 +556,32 @@ my $sendRecvCleanup = sub {
...
             next if ($backupSet->{recursive} eq 'on' && $srcDataSet eq $backupSet->{src});
 
+            $self->zLog->debug('checking to clean up snapshots of source '. $srcDataSet);
+
             @snapshots = @{$self->zZfs->listSnapshots($srcDataSet, $backupSet->{snapCleanFilter})};
             $toDestroy = $self->zTime->getSnapshotsToDestroy(\@snapshots,
                          $backupSet->{srcPlanHash}, $backupSet->{tsformat}, $timeStamp);
 
+            for (@{$toDestroy}) {
+                my $snapname = $_;
```
Comment (oetiker) [inline lib/ZnapZend.pm:575]: is this to make sure that the content of the $toDestroy array does not get changed ?

File: lib/ZnapZend.pm [hunk 8]
```diff
@@ -474,11 +479,24 @@ my $sendRecvCleanup = sub {
...
             # and are all committed at once.
+            $self->zLog->debug('checking to clean up snapshots recursively from source '. $backupSet->{src});
 
             @snapshots = @{$self->zZfs->listSnapshots($backupSet->{src}, $backupSet->{snapCleanFilter})};
             $toDestroy = $self->zTime->getSnapshotsToDestroy(\@snapshots,
                          $backupSet->{srcPlanHash}, $backupSet->{tsformat}, $timeStamp);
 
+            # Save the names we have seen, to not revisit them below for children
+            for (@snapshots) {
+                my $snapname = $_;
```
Comment (oetiker) [inline lib/ZnapZend.pm:490]: is this to make sure that the content of the snapshots array does not get changed ?

File: lib/ZnapZend.pm [hunk 9]
```diff
@@ -474,11 +479,24 @@ my $sendRecvCleanup = sub {
...
+                $snapname =~ s/^.*\@//;
+                $snapnamesRecursive{$snapname} = 1;
+            }
+            for (@{$toDestroy}) {
+                my $snapname = $_;
```
Comment (oetiker) [inline lib/ZnapZend.pm:495]: is this to make sure that the toDestroy array does not get changed ?

File: lib/ZnapZend.pm (hunk 7 above)
Comment (oetiker) [reply lib/ZnapZend.pm:575]: maybe having a function that takes care of extracting the snapshot names, and returning an array would make sense here ?

File: lib/ZnapZend/Config.pm [hunk 10]
```diff
@@ -12,7 +12,11 @@ has zfsGetType => sub { 0 };
 has noaction => sub { 0 };
 has rootExec => sub { q{} };
 has timeWarp => sub { undef };
-has zLog => sub { Mojo::Exception->throw('zLog must be specified at creation time!') };
+has zLog            => sub {
```
Comment (oetiker) [inline lib/ZnapZend/Config.pm:15]: spacing

File: lib/ZnapZend/ZFS.pm [hunk 11]
```diff
@@ -948,45 +1038,169 @@ sub getSnapshotProperties {
...
     }
+    my $numProps = keys %properties;
+
+    if ($self->debug) {
+        if ($numProps > 0) {
+            $self->zLog->debug("=== getSnapshotProperties(): GOT $inhMode (#$inherit) properties of $snapshot : " .Dumper(\%properties) );
+        }
+    }
+
+    if ($inherit == 2 || $inherit == 3) {
```
Comment (oetiker) [inline lib/ZnapZend/ZFS.pm:1110]: I find such long multi level if/then constructs a bit hard to follow ... how about returning the `\%properties if $inherit < 2` here to loose a level ?

File: lib/ZnapZend.pm (hunk 7 above)
Comment (jimklimov) [reply lib/ZnapZend.pm:575]: good point, this bit of code happens a few times too many :)

File: lib/ZnapZend.pm (hunk 8 above)
Comment (jimklimov) [reply lib/ZnapZend.pm:490]: yes, I found that with `for my $snapname (@toDestroy)` at least, regexing the `$snapname` in the next line did change the array entries from `dataset@snapname` to `snapname`.

File: lib/ZnapZend/ZFS.pm (hunk 11 above)
Comment (jimklimov) [reply lib/ZnapZend/ZFS.pm:1110]: I originally thought to make it a bit mask, then settled for a number of magic values with specific meanings (commented above) that are a bit faster to compare than more meaningful strings or further variables - and we pass this codepath a lot when we do :\

Comment (oetiker) [reply lib/ZnapZend/ZFS.pm:1110]: oh I did not realize that ... please don't :) use a hash with speaking  keys for the features you want enabled ...

File: lib/ZnapZend.pm (hunk 7 above)
Comment (jimklimov) [reply lib/ZnapZend.pm:575]: Although not sure on second thought: I tracked twice the snapshots we considered (in filtered `@snapshots` and selection `@toDestroy`) with different values... I guess there was an idea to differentiate the two though not sure now, probably it did not get into code and popped out of memory at the moment. So we'd need then to merge the two resulting arrays also, instead of now just writing into one. Or pass several to one routine...

File: lib/ZnapZend/ZFS.pm (hunk 11 above)
Comment (jimklimov) [reply lib/ZnapZend/ZFS.pm:1110]: I think I get the idea but did not find an example to rip off, do you have some handy please?

**Comment by jimklimov** [comment]
Another TODO (now or in another PR to not stall this one): if the idea with not-revisiting snapshot cleanup for tags we have considered in recursive mode seems okay, it can be replicated into destination clean-ups too.

File: lib/ZnapZend/ZFS.pm (hunk 11 above)
Comment (oetiker) [reply lib/ZnapZend/ZFS.pm:1110]:
```perl
my $inherit = {
    all => 1,
    greens => 0,
    reds => 1
};


if ($inherit->{all} or $inherit->{greens}){ .... 

}
```

File: lib/ZnapZend.pm (hunk 7 above)
Comment (oetiker) [reply lib/ZnapZend.pm:581]: What I mean is that instead of walking the array and running the regexp to extract the snapshot names you could create a function which does that

```perl 
sub extractSnapshotNames($array){
      my @ret;
      for (@$array){
           /\@(.+)/ and push @ret, $1;
     }
     return \@ret;
}
```

File: lib/ZnapZend.pm (hunk 8 above)
Comment (oetiker) [reply lib/ZnapZend.pm:496]: this is a feature :) see my proposed function above

File: lib/ZnapZend/ZFS.pm (hunk 11 above)
Comment (jimklimov) [reply lib/ZnapZend/ZFS.pm:1110]: Ah, just so? I thought there was something computationally cheaper, like C enums. Thanks, will try.

File: lib/ZnapZend.pm (hunk 7 above)
Comment (jimklimov) [reply lib/ZnapZend.pm:581]: Thanks, I'll see if a usecase remains for tracking that as a hash with various levels, or an array is okay.

Comment (jimklimov) [reply lib/ZnapZend.pm:581]: I guess the hash usage originated with another sanity check, to make sure values are unique just in case. And since we had a hash, I decided to keep it, e.g. to have speedy lookups into the keys (is this snapname in the list?)...

Comment (jimklimov) [reply lib/ZnapZend.pm]: Done, pushed, tested

File: lib/ZnapZend/ZFS.pm (hunk 11 above)
Comment (jimklimov) [reply lib/ZnapZend/ZFS.pm]: I suppose my fix is not what you wanted (ended up like an enum), but cursorily seems to work and steps away from arcane numbers to human names for them, but still uses the number underneath. Might revise later to have it as a flag map per your example, but I think it would be a refactor irrelevant to functionality at the moment :)

File: lib/ZnapZend/ZFS.pm [hunk 12]
```diff
@@ -213,6 +227,46 @@ sub listSnapshots {
     return \@snapshots;
 }
 
+sub extractSnapshotNames {
+    # Routines like listSnapshots() operate with fully qualified
+    # "dataset@snapname" strings as returned by ZFS. For some data
+    # matching we need to compare just the "snapname" lists.
+    my $self = shift;
+    my ($array) = @_; # Note: call with extractSnapshotNames(\@arrVarName) !
```
[the remark this answers is not in the document]
Comment (jimklimov) [reply lib/ZnapZend/ZFS.pm:235]: I tried that originally, seems this was the syntax used in other routines, but seems it was about as messy (especially with need to pass `\@arrname` vs `@arrname`. 

Neither approach handled arguments coming from `qw(word list)` for example; I hoped this variant to catch cases where someone passes multiple single string arguments by mistake or somehow else, but must have got perl syntax wrong somewhere.

[the remark this answers is not in the document]
Comment (jimklimov) [reply lib/ZnapZend/ZFS.pm:235]: Re-checked to be sure, behaves the same in the three use-cases commented after the routine :)

**Comment by jimklimov** [comment]
Hm, seems like we say the snapshots seen in recursive mode would not be revisited in toDestroy list, but then the list still contains them. Gotta check that this grep works...

**Comment by jimklimov** [comment]
Looks better now, almost no snaps to clean in the child phase because they were looked at in the recursive phase and decidedly left to stick around.

**Comment by jimklimov** [comment]
Interesting, I think I over-engineered my worries for now: in manual experiments I was using `zfs send -R -I ... | zfs recv` for quicker wholesale replications -- something I intended or asked long time ago to use in `znapzend` for first stage sendings (like recursive clean-ups, that might not need following by child-by-child cleanups if all went well), so I assumed somehow this was the mode it used already - and then it would only mark the topmost parent dataset of a replication run with `dst_X_synced` flags in `sendRecvSnapshot()`.

Well, we're saved by the bell: that mode is not in codebase yet, so every dataset in a subtree is replicated individually, marked as synced individually, and iteration to get the snapshot properties involved does not go far and cost much even without us making some caching and tree-pruning :)

**Comment by jimklimov** [comment]
Applied a few patches from #497 that do fix issues and would conflict the remaining PR after merging one to master first.

File: lib/ZnapZend/ZFS.pm [hunk 13]
```diff
@@ -12,12 +12,12 @@ use Data::Dumper;
...
 ### properties inherited from higher level datasets; but for
 ### snapshots it only returns the same - not from higher snaps.
-my %inheritLevels = {
+my %inheritLevels = (
     local_only => 0,
     local_zfsinherit => 1,
     local_recurseparent => 2,
     local_recurseparent_zfsinherit => 3,
-};
+);
```
Comment (oetiker) [inline lib/ZnapZend/ZFS.pm:20]: as you mentioned :) sort of semi satisfying this way ...

Comment (jimklimov) [reply lib/ZnapZend/ZFS.pm:20]: This PR is stalling and growing too large to review well already :( Can we leave such refinements for later iterations? I believe after the "proper fix" functionality would be the same, maintainability (key words instead of magic numbers) too, so that classifies as a refactor and can be separated :)

Comment (oetiker) [reply lib/ZnapZend/ZFS.pm:20]: Sure if you would like to follow up with another one to fix this properly ... I am fine with merging this one now. 

The reason I am trying to catch these things is that they are non-ideomatic and tend to make my life harder when I have to figure out some odd behavior some time down the road. As I tend to be the one stuck with these projects in the long run.

**Comment by oetiker** [comment]
thanks for your contribution
